engines/cryomni3d/versailles/logic.cpp

```diff
@@ -0,0 +1,4650 @@
+			            &CryOmni3DEngine_Versailles::img_34174f);
+			fimg->changeCallback(functor);
+			break;
+		}
+	}
+}
+
+IMG_CB(34174f) {
+	// Safe is empty
+	fimg->load("43X3_45.GIF");
+	while (1) {
+		fimg->manage();
+		if (fimg->_exit || fimg->_zoneLow) {
+			fimg->_exit = true;
+			break;
+		}
+		if (fimg->_zoneUse) {
+			// Close safe
+			ZonFixedImage::CallbackFunctor *functor =
+			    new Common::Functor1Mem<ZonFixedImage *, void, CryOmni3DEngine_Versailles>(this,
+			            &CryOmni3DEngine_Versailles::img_34174b);
+			fimg->changeCallback(functor);
+			break;
+		}
+	}
+}
+
+bool CryOmni3DEngine_Versailles::handleSafe(ZonFixedImage *fimg) {
+	bool success = false;
+	Common::RandomSource rnd("VersaillesSafe");
+	Graphics::Surface bmpDigits[10];
+	unsigned char safeDigits[kSafeDigitsCount];
+	Graphics::ManagedSurface tempSurf;
+
+	loadBMPs("coff_%02d.bmp", bmpDigits, 10);
+	for (unsigned int i = 0; i < kSafeDigitsCount; i++) {
+		safeDigits[i] = rnd.getRandomNumber(9);
+	}
+
+	fimg->load("43x3_cof.GIF");
+	const Graphics::Surface *fimgSurface = fimg->surface();
+	tempSurf.create(fimgSurface->w, fimgSurface->h, fimgSurface->format);
+	tempSurf.blitFrom(*fimgSurface);
+	drawSafeDigits(tempSurf, bmpDigits, safeDigits);
+	fimg->updateSurface(&tempSurf.rawSurface());
+
+	while (1) {
+		fimg->manage();
+		if (fimg->_exit || fimg->_zoneLow) {
+			break;
+		}
+		if (fimg->_zoneUse) {
+			if (fimg->_currentZone == 15) {
+				// Safe handle
+
+				// Animate handle
+				playInGameVideo("43x3_poi");
+				// Force reload of the place
+				if (_nextPlaceId == -1u) {
+					_nextPlaceId = _currentPlaceId;
+				}
+
+				// Redraw our safe image
+				fimg->display();
+
+				if (checkSafeDigits(safeDigits)) {
+					success = true;
+					break;
+				}
+			} else if (fimg->_currentZone < kSafeDigitsCount) {
+				// Safe digit
+				safeDigits[fimg->_currentZone] = (safeDigits[fimg->_currentZone] + 1) % 10;
+				// Reset the surface and redraw digits on it
+				tempSurf.blitFrom(*fimgSurface);
+				drawSafeDigits(tempSurf, bmpDigits, safeDigits);
+				fimg->updateSurface(&tempSurf.rawSurface());
+
+				waitMouseRelease();
+			}
+		}
+	}
+
+	for (unsigned int i = 0; i < 10; i++) {
+		bmpDigits[i].free();
+	}
+	return success;
+}
+
+const unsigned short CryOmni3DEngine_Versailles::kSafeDigitsX[] = { 267, 318, 370, 421 };
+const unsigned short CryOmni3DEngine_Versailles::kSafeDigitsY[] = { 148, 230, 311 };
+
+void CryOmni3DEngine_Versailles::drawSafeDigits(Graphics::ManagedSurface &surface,
+        const Graphics::Surface(&bmpDigits)[10], const unsigned char (&safeDigits)[kSafeDigitsCount]) {
+	for (unsigned int i = 0; i < ARRAYSIZE(safeDigits); i++) {
+		const Graphics::Surface &digit = bmpDigits[safeDigits[i]];
+		Common::Point dst(kSafeDigitsX[i % 4], kSafeDigitsY[i / 4]);
+		surface.transBlitFrom(digit, dst);
+	}
+}
+
+const char *CryOmni3DEngine_Versailles::kSafeDates[] = { "1643", "1668", "1674" };
```

Comment:
This is named incorrectly, it is not a constant but an array

Reply:
It seems there are many engines using kConstant names for arrays. Should I really change it?

Reply:
Any example? In the end, since the usage of this array is pretty local, it may stay as is.

Reply:
Using : `grep -R ' k[A-Z].*\[\]' ./engines/` I found that tony, pegasus, sci, gob, draci, dreawweb, supernova, gnap, neverhood, bladerunner, illusions and cryo have some. bbvs has a lot of them.